### PR TITLE
Added Behat Tests for Language Domain

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Added Behat Tests for Language Domain
| Type?             | improvement
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | N/A
| How to test?      | Github Actions is :green_circle: 
| Possible impacts? | N/A

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26362)

### CODE REVIEW

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    }
+
+    /**
+     * @Then the language with ISOCode :isoCode should exist
+     *
+     * @param string $isoCode
+     */
+    public function checkLanguageExists(string $isoCode): void
+    {
+        $languageId = Language::getIdByIso($isoCode, true);
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: I think this variable is useless, you can directly put this in the assertNotNull method :)

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    }
+
+    /**
+     * @Then the language with ISOCode :isoCode shouldn't exist
+     *
+     * @param string $isoCode
+     */
+    public function checkLanguageNotExists(string $isoCode): void
+    {
+        $languageId = Language::getIdByIso($isoCode, true);
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: I think this variable is useless, you can directly put this in the assertNull method :)

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+     * @param string $statusVerb
+     * @param string $isoCode
+     */
+    public function setStatusLanguage(string $statusVerb, string $isoCode): void
+    {
+        $languageId = Language::getIdByIso($isoCode, true);
+        if (!is_numeric($languageId)) {
+            throw new RuntimeException(sprintf('No language found with the ISOCode "%s"', $isoCode));
+        }
+        if (!in_array($statusVerb, ['enable', 'disable'])) {
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: You use this array many times, maybe you can put this in a constant or do a private method that does this check

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    And the language with ISOCode "ag" should have the following details:
+      | name            | Español AR (Spanish) |
+      | isoCode         | ag                   |
+      | tagIETF         | es-ar                |
+      | shortDateFormat | Y-m-d                |
+      | fullDateFormat  | Y-m-d H:i:s          |
+      | isRtl           | 0                    |
+      | isActive        | 1                    |
+      | shop            | shop1                |
+    When I delete the language with ISOCode "ag"
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:49]: I don't understand why you delete, in the "add" test?

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      | isRtl           | 0                 |
+      | isActive        | 1                 |
+      | shop            | shop1             |
+
+  Scenario: Delete language
+    Given the language with ISOCode "fr" should exist
+    When I delete the language with ISOCode "fr"
+    Then I should get no error
+    And the language with ISOCode "fr" shouldn't exist
+    When I add a new language with the following details:
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Maybe this test should have its scenario, to separate tests with error from those with success

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    When language with iso code "en" is the default one
+
+  Scenario: Bulk Delete
+    Given the language with ISOCode "fr" should exist
+    And the language with ISOCode "gb" should exist
+    When I bulk delete languages with ISOCode "fr,gb"
+    Then I should get no error
+    And the language with ISOCode "fr" shouldn't exist
+    And the language with ISOCode "gb" shouldn't exist
+    When I add a new language with the following details:
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: I think you have to split the scenario, to be more readable. Your test is big, IMO

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      | isRtl           | 0                    |
+      | isActive        | 1                    |
+      | shop            | shop1                |
+    Then I should get no error
+    And the language with ISOCode "ag" should be enabled
+    When I disable the language with ISOCode "ag"
+    Then the language with ISOCode "ag" should be disabled
+    When I enable the language with ISOCode "ag"
+    And the language with ISOCode "ag" should be enabled
+    When language with iso code "ag" is the default one
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:120]: Maybe this test should have its scenario, to separate tests with error from those with success

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  Scenario: Bulk Toggle Status
+    Given the language with ISOCode "fr" should be enabled
+    And the language with ISOCode "ag" should be enabled
+    When I bulk disable languages with ISOCode "ag,fr"
+    Then I should get no error
+    And   the language with ISOCode "fr" should be disabled
+    And the language with ISOCode "ag" should be disabled
+    When I bulk enable languages with ISOCode "ag,fr"
+    Then I should get no error
+    And   the language with ISOCode "fr" should be enabled
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Typo here

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    Then I should get an error that a default language can't be disabled
+    ## Reset
+    When language with iso code "en" is the default one
+
+  Scenario: Bulk Toggle Status
+    Given the language with ISOCode "fr" should be enabled
+    And the language with ISOCode "ag" should be enabled
+    When I bulk disable languages with ISOCode "ag,fr"
+    Then I should get no error
+    And   the language with ISOCode "fr" should be disabled
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Type here

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    And the language with ISOCode "ag" should be disabled
+    When I bulk enable languages with ISOCode "ag,fr"
```
Comment (tegbessou) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:134]: Maybe this test should have its scenario, to separate tests with error from those with success

**Review by tegbessou — COMMENTED** [review]
Thank you for your @Progi1984, I do some returns

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php (hunk 1 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: Fixed

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php (hunk 2 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: Fixed

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php (hunk 3 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: Removed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 4 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:49]: It's a reset (comment added)

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 5 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 6 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 7 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:120]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 8 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 9 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 10 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:134]: Fixed

**Review by matthieu-rolland — DISMISSED** [review]

**Review by tegbessou — DISMISSED** [review]
Thank you for your contribution! IMO all is good!

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+
+    /**
+     * @Then the language with ISOCode :isoCode should exist
+     *
+     * @param string $isoCode
+     */
+    public function checkLanguageExists(string $isoCode): void
+    {
+        Assert::assertNotNull(
+            Language::getIdByIso($isoCode, true),
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: You shouldn't rely on legacy code for this, it's better to use the query `GetLanguageForEditing` catch the error and assert that a "NotFoundException" has been thrown

It has two advantages:
- you can test that the query handler is able to detect a non existing language (base on a removed ID)
- you on't have any more problems related to the cache from `Language::getIdByIso`

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+
+    /**
+     * @Then the language with ISOCode :isoCode shouldn't exist
+     *
+     * @param string $isoCode
+     */
+    public function checkLanguageNotExists(string $isoCode): void
+    {
+        Assert::assertNull(
+            Language::getIdByIso($isoCode, true),
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php]: Same as the check exists, you should rely on the query here as well

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    When I delete the language with ISOCode "fr"
+    Then I should get no error
+    And the language with ISOCode "fr" shouldn't exist
+
+  Scenario: Delete a default language
+    Given the language with ISOCode "fr" should exist
+    When language with iso code "fr" is the default one
+    When I delete the language with ISOCode "fr"
+    Then I should get an error that a default language can't be deleted
+    ## Reset
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature]:
```suggestion
    ## Reset default language back to english
```

I didn't understand at first

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    Given the language with ISOCode "fr" should exist
+    When I delete the language with ISOCode "fr"
+    Then I should get no error
+    And the language with ISOCode "fr" shouldn't exist
+
+  Scenario: Delete a default language
+    Given the language with ISOCode "fr" should exist
+    When language with iso code "fr" is the default one
+    When I delete the language with ISOCode "fr"
+    Then I should get an error that a default language can't be deleted
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:80]: You should also add a test to ensure that the fr language still exists

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+  Scenario: Bulk Delete
+    Given the language with ISOCode "fr" should exist
+    And the language with ISOCode "gb" should exist
+    When I bulk delete languages with ISOCode "fr,gb"
+    Then I should get no error
+    And the language with ISOCode "fr" shouldn't exist
+    And the language with ISOCode "gb" shouldn't exist
+
+  Scenario: Bulk Delete (with a default one)
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:91]: I get the idea here, but it's a bit weird that the behaviour is different depending on the order of the input...
I wonder if the check should be performed before any languages is removed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    And the language with ISOCode "ag" should be enabled
+    When I disable the language with ISOCode "ag"
+    Then the language with ISOCode "ag" should be disabled
+    When I enable the language with ISOCode "ag"
+    And the language with ISOCode "ag" should be enabled
+
+  Scenario: Toggle the status of the default language
+    When language with iso code "ag" is the default one
+    And I disable the language with ISOCode "ag"
+    Then I should get an error that a default language can't be disabled
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:122]: Add a test to ensure fr is still enabled

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+  Scenario: Bulk Toggle Status (with a default one)
+    When language with iso code "ag" is the default one
+    When I bulk disable languages with ISOCode "ag,fr"
+    Then I should get an error that a default language can't be disabled
+    And the language with ISOCode "fr" should be enabled
+    And the language with ISOCode "ag" should be enabled
+    When I bulk disable languages with ISOCode "fr,ag"
+    Then I should get an error that a default language can't be disabled
+    And the language with ISOCode "fr" should be disabled
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Language/language.feature:147]: Same suggestion as the bulk delete It seems a bit unstable that the order has impact on the result

It should be either every time the update is performed and the exception is raised (delay the exception throwing) OR the update is never done and the exception is raised (check and throw exception before any update)

But not a mix of different behaviour depending on the order of inputs

**Review by jolelievre — COMMENTED** [review]
Some additional comments

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 16 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:122]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 14 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:80]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 15 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:91]: I think you should create and issue for a product decision.

In this case, we can merge this PR without QA.

File: tests/Integration/Behaviour/Features/Scenario/Language/language.feature (hunk 17 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Language/language.feature:147]: I think you should create and issue for a product decision.

In this case, we can merge this PR without QA.

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+        foreach ($isoCodes as $isoCode) {
+            $languageIds[] = SharedStorage::getStorage()->get($isoCode);
+        }
+
+        try {
+            $this->cleanLastException();
+            $this->getCommandBus()->handle(new BulkDeleteLanguagesCommand($languageIds));
+
+            // Important to clean this cache or Language::getIdByIso still returns stored value for next adding
+            Cache::clean('Language::getIdByIso_*');
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php:177]: Maybe you can try to remove this now? If it doesn't work don't worry about it

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    /**
+     * @Then the language with ISOCode :isoCode should exist
+     *
+     * @param string $isoCode
+     */
+    public function checkLanguageExists(string $isoCode): void
+    {
+        $this->getLanguage($isoCode);
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php:264]:
```suggestion
        $this->assertNull($this->getLanguage($isoCode));
```

Just a suggestion, the exception alone is enough to validate the behaviour

**Review by jolelievre — APPROVED** [review]
Just a couple suggestions, but not blocking

**Comment by jolelievre** [comment]
I created the issue regarding the bulk actions inconsistency https://github.com/PrestaShop/PrestaShop/issues/26453

@MatShir or @marionf maybe you can have a look at it and tell us what you think should be the expected behaviour

File: tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php (hunk 18 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/LanguageFeatureContext.php:177]: @jolelievre I try it. And we need it

**Review by matthieu-rolland — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
thank you @Progi1984

**Pull request merged**